srml/staking/src/lib.rs

```diff
@@ -585,8 +585,18 @@ decl_storage! {
 					};
 				}
 
+				// Make sure the genesis state of the session module is consistent
+				// with the genesis state of the staking module.
 				if let (_, Some(validators)) = <Module<T>>::select_validators() {
-					<session::Validators<T>>::put(&validators);
+					let keys = validators.iter().map(|validator| {
+						(
+							validator.to_owned(),
+							<session::Module<T>>::next_key_for(validator)
+								.unwrap_or_default()
+						)
+					}).collect::<Vec<_>>();
+					<session::Validators<T>>::put(validators);
+					<session::QueuedKeys<T>>::put(keys);
```

Comment:
This will clearly blow up at some point. You set `session::QueuedKeys` in the session `GenesisConfig` and here. Currently that works, as `staking` is coming after `session` in the `construct_runtime!` in our runtime, but anyone reversing this order will get different results.

Reply:
I agree, but that's how it was done before already...

Reply:
Hmm okay :( 
Could you then use `extra_genesis_build` in `session` and put an assert, that the keys are not already initialized?

Reply:
I don't think that's a good idea, since the session module can be used without the staking module. That makes assumptions that are broken at least in the session tests.

Reply:
If the session modules make sure that while generating the genesis storage, the keys are not already set, how does this break tests? The genesis config wants to initialize these storage items and if they are already present, something is incorrect. 
We could combine this check with a check that someone really wants to set some keys, because if the genesis config does not want to set keys, we don't need to check if some keys are already set.

Reply:
Why should it be the responsibility of `staking` to set the `session` genesis set?

Reply:
Maybe I'm missunderstanding, but this doesn't work...
```diff
diff --git a/srml/session/Cargo.toml b/srml/session/Cargo.toml
index 036ac39e..a13fe4d6 100644
--- a/srml/session/Cargo.toml
+++ b/srml/session/Cargo.toml
@@ -9,6 +9,7 @@ serde = { version = "1.0", optional = true }
 safe-mix = { version = "1.0", default-features = false}
 parity-codec = { version = "4.1.1", default-features = false, features = ["derive"] }
 rstd = { package = "sr-std", path = "../../core/sr-std", default-features = false }
+runtime_io = { package = "sr-io", path = "../../core/sr-io", default-features = false }
 primitives = { package = "sr-primitives", path = "../../core/sr-primitives", default-features = false }
 srml-support = { path = "../support", default-features = false }
 system = { package = "srml-system", path = "../system", default-features = false }
@@ -16,7 +17,6 @@ timestamp = { package = "srml-timestamp", path = "../timestamp", default-feature
 
 [dev-dependencies]
 substrate-primitives = { path = "../../core/primitives" }
-runtime_io = { package = "sr-io", path = "../../core/sr-io" }
 lazy_static = "1.0"
 
 [features]
diff --git a/srml/session/src/lib.rs b/srml/session/src/lib.rs
index a3aeff1f..aedcfaf7 100644
--- a/srml/session/src/lib.rs
+++ b/srml/session/src/lib.rs
@@ -118,6 +118,8 @@
 use rstd::{prelude::*, marker::PhantomData, ops::Rem};
 #[cfg(not(feature = "std"))]
 use rstd::alloc::borrow::ToOwned;
+#[cfg(feature = "std")]
+use runtime_io::with_storage;
 use parity_codec::Decode;
 use primitives::traits::{Zero, Saturating, Member, OpaqueKeys};
 use srml_support::{
@@ -265,6 +267,16 @@ decl_storage! {
        }
        add_extra_genesis {
                config(keys): Vec<(T::AccountId, T::Keys)>;
+               build(|
+                         storage: &mut primitives::StorageOverlay,
+                         _: &mut primitives::ChildrenStorageOverlay,
+                         _: &GenesisConfig<T>
+               | {
+                       with_storage(storage, || {
+                               assert_eq!(<Validators<T>>::get(), vec![]);
+                               assert_eq!(<QueuedKeys<T>>::get(), vec![]);
+                       });
+               });
        }
 }
```

Reply:
IMO setting the genesis values correctly for `session` should be the responsibility of the genesis configuration, not to be done in `staking`. Or if it is, then it should be done via a trait parameter to the `session` module.

Reply:
Do you want to turn it into a compile time option like this?

```diff
diff --git a/srml/session/src/lib.rs b/srml/session/src/lib.rs
index aedcfaf7..f0af6066 100644
--- a/srml/session/src/lib.rs
+++ b/srml/session/src/lib.rs
@@ -222,6 +222,9 @@ pub trait Trait: system::Trait {
 
        /// The keys.
        type Keys: OpaqueKeys + Member + Parameter + Default;
+
+       /// Initial keys.
+       type InitialKeys: Get<Vec<(Self::AccountId, Self::Keys)>>;
 }
 
 type OpaqueKey = Vec<u8>;
@@ -229,8 +232,8 @@ type OpaqueKey = Vec<u8>;
 decl_storage! {
        trait Store for Module<T: Trait> as Session {
                /// The current set of validators.
-               pub Validators get(validators) build(|config: &GenesisConfig<T>| {
-                       config.keys.iter().map(|(validator, _)| validator.to_owned()).collect()
+               pub Validators get(validators) build(|_| {
+                       T::InitialKeys::get().iter().map(|(validator, _)| validator.to_owned()).collect()
                }): Vec<T::AccountId>;
 
                /// Current index of the session.
@@ -242,8 +245,8 @@ decl_storage! {
                /// The next key to be used for a given validator. At the end of the session, they
                /// will be moved into the `QueuedKeys` and so changes here will not take effect for
                /// at least one whole session.
-               NextKeyFor get(next_key_for) build(|config: &GenesisConfig<T>| {
-                       config.keys.clone()
+               NextKeyFor get(next_key_for) build(|_| {
+                       T::InitialKeys::get()
                }): map T::AccountId => Option<T::Keys>;
 
                /// Queued keys changed.
@@ -251,33 +254,20 @@ decl_storage! {
 
                /// The queued keys for the next session. When the next session begins, these keys
                /// will be used to determine the validator's session keys.
-               pub QueuedKeys get(queued_keys) build(|config: &GenesisConfig<T>| {
-                       config.keys.clone()
+               pub QueuedKeys get(queued_keys) build(|_| {
+                       T::InitialKeys::get()
                }): Vec<(T::AccountId, T::Keys)>;
 
                /// The keys that are currently active.
-               Active build(|config: &GenesisConfig<T>| {
+               Active build(|_| {
                        (0..T::Keys::count()).map(|i| (
                                i as u32,
-                               config.keys.iter()
+                               T::InitialKeys::get().iter()
                                        .map(|x| x.1.get_raw(i).to_vec())
                                        .collect::<Vec<OpaqueKey>>(),
                        )).collect::<Vec<(u32, Vec<OpaqueKey>)>>()
                }): map u32 => Vec<OpaqueKey>;
        }
-       add_extra_genesis {
-               config(keys): Vec<(T::AccountId, T::Keys)>;
-               build(|
-                         storage: &mut primitives::StorageOverlay,
-                         _: &mut primitives::ChildrenStorageOverlay,
-                         _: &GenesisConfig<T>
-               | {
-                       with_storage(storage, || {
-                               assert_eq!(<Validators<T>>::get(), vec![]);
-                               assert_eq!(<QueuedKeys<T>>::get(), vec![]);
-                       });
-               });
-       }
 }
 
 decl_event!(
```

Reply:
No.
There is clearly a spectrum between "there should never be any genesis keys from the config" and "spooky action at a distance".

Reply:
As I'm clearly in favor of what @rphmeier is saying.

Applying the following patch on top of your last commit:
```
diff --git a/srml/session/src/lib.rs b/srml/session/src/lib.rs
index aedcfaf7..c05e6019 100644
--- a/srml/session/src/lib.rs
+++ b/srml/session/src/lib.rs
@@ -251,9 +251,7 @@ decl_storage! {
 
                /// The queued keys for the next session. When the next session begins, these keys
                /// will be used to determine the validator's session keys.
-               pub QueuedKeys get(queued_keys) build(|config: &GenesisConfig<T>| {
-                       config.keys.clone()
-               }): Vec<(T::AccountId, T::Keys)>;
+               pub QueuedKeys get(queued_keys): Vec<(T::AccountId, T::Keys)>;
 
                /// The keys that are currently active.
                Active build(|config: &GenesisConfig<T>| {
@@ -270,11 +268,11 @@ decl_storage! {
                build(|
                          storage: &mut primitives::StorageOverlay,
                          _: &mut primitives::ChildrenStorageOverlay,
-                         _: &GenesisConfig<T>
+                         config: &GenesisConfig<T>
                | {
                        with_storage(storage, || {
-                               assert_eq!(<Validators<T>>::get(), vec![]);
                                assert_eq!(<QueuedKeys<T>>::get(), vec![]);
+                           <QueuedKeys<T>>::put(&config.keys);
                        });
                });
        }
```

Would get you to what I meant. (clearly missing the same for `Validators`.)

Nevertheless, we should come to a solution without modifying the `session` genesis config from outside the `session` module.